experimental/dds/tree2/src/class-tree/integerIndexable.ts

```diff
@@ -0,0 +1,123 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assert } from "@fluidframework/core-utils";
+
+// Note:
+// Currently this file is unused, though it could serve as the implementation of indexing for Lists when class and simple trees are merged.
+
+/**
+ * Info integerIndexable needs to work.
+ */
+export interface IndexLookup<T> {
+	read(index: number): T;
+	write(index: number, value: T): boolean;
+	readonly length: number;
+}
+
+/**
+ * Class which can indexed with integers from 0 to length -1 like arrays.
+ * Unlike with arrays, this class asserts for out of bounds indexing on both read and write.
+ *
+ * @privateRemarks
+ * This class works by returning a proxy from its constructor which
+ * causes the derived class to get the proxy as its "this" object.
+ */
+export abstract class IntegerIndexable<in out T> {
+	[x: number]: T;
+
+	protected abstract read(index: number): T;
+	protected abstract write(index: number, value: T): boolean;
+
+	public constructor(public length: number) {
+		return integerIndexable(this, this as unknown as IndexLookup<T>);
+	}
+}
+
+export function integerIndexable<T, TTarget extends object>(
+	outerTarget: TTarget,
+	provider: IndexLookup<T>,
+): {
+	[x: number]: T;
+} & TTarget {
+	function isValidIndex(index: number): boolean {
+		return index >= 0 && index < provider.length;
```

Comment:
indexing and the "at" function are different. When given a negative index, JavaScript arrays return undefined (just like they do for any other out of bounds index), and typescript types that value as the member type, resulting in a value with the wrong type at runtime.

Personally, I think it makes more sense to throw than return a value that is undefined when the API says you don't. Regardless none of this has to do with "at" which does its special negative index thing. If implementing "at" ontop of this, it would need to do its own negative index logic before calling this.